NPE if parent POM and micronaut.version differ

If NPE if parent POM and micronaut.version differ an exception occurs:

```
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.List.contains(Object)" because the return value of "java.util.Map.get(Object)" is null
    at io.micronaut.build.DockerNativeMojo.checkGraalVm (DockerNativeMojo.java:102)
    at io.micronaut.build.DockerNativeMojo.execute (DockerNativeMojo.java:65)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apach
```

--- a/pom.xml
+++ b/pom.xml
@@ -80,7 +80,7 @@
         <native-maven-plugin.version>0.9.11</native-maven-plugin.version>
         <micronaut.version>3.5.0</micronaut.version>
         <micronaut.aot.version>1.1.0</micronaut.aot.version>
-        <it.micronaut.version>3.5.0-SNAPSHOT</it.micronaut.version>
+        <it.micronaut.version>3.5.1-SNAPSHOT</it.micronaut.version>
         <sonar.host.url>https://sonarcloud.io</sonar.host.url>
         <sonar.organization>micronaut-projects</sonar.organization>
         <sonar.projectKey>micronaut-projects_micronaut-maven-plugin</sonar.projectKey>

--- a/src/it/package-docker-native-different-parent-version/invoker.properties
+++ b/src/it/package-docker-native-different-parent-version/invoker.properties
@@ -1,0 +1,4 @@
+invoker.goals = -ntp package -Dpackaging=docker-native
+invoker.profiles = graalvm
+invoker.os.family = !windows
+invoker.buildResult = failure

--- a/src/it/package-docker-native-different-parent-version/pom.xml
+++ b/src/it/package-docker-native-different-parent-version/pom.xml
@@ -1,0 +1,137 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>io.micronaut.build.examples</groupId>
+  <artifactId>package-docker-native-different-parent-version</artifactId>
+  <version>0.1</version>
+  <packaging>${packaging}</packaging>
+
+  <parent>
+    <groupId>io.micronaut</groupId>
+    <artifactId>micronaut-parent</artifactId>
+    <version>3.4.0</version>
+  </parent>
+
+  <properties>
+    <jdk.version>8</jdk.version>
+    <release.version>8</release.version>
+    <micronaut.version>@it.micronaut.version@</micronaut.version>
+    <micronaut-maven-plugin.version>@project.version@</micronaut-maven-plugin.version>
+    <exec.mainClass>io.micronaut.build.examples.Application</exec.mainClass>
+    <packaging>jar</packaging>
+    <micronaut.runtime>netty</micronaut.runtime>
+  </properties>
+
+  <repositories>
+    <repository>
+      <id>central</id>
+      <url>https://repo.maven.apache.org/maven2</url>
+    </repository>
+    <repository>
+      <id>sonatype-snapshots</id>
+      <url>https://s01.oss.sonatype.org/content/repositories/snapshots/</url>
+      <snapshots>
+        <enabled>true</enabled>
+      </snapshots>
+      <releases>
+        <enabled>false</enabled>
+      </releases>
+    </repository>
+  </repositories>
+
+  <dependencies>
+    <dependency>
+      <groupId>io.micronaut</groupId>
+      <artifactId>micronaut-inject</artifactId>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.micronaut</groupId>
+      <artifactId>micronaut-validation</artifactId>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-api</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.micronaut.test</groupId>
+      <artifactId>micronaut-test-junit5</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.micronaut</groupId>
+      <artifactId>micronaut-http-client</artifactId>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.micronaut</groupId>
+      <artifactId>micronaut-http-server-netty</artifactId>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.micronaut</groupId>
+      <artifactId>micronaut-jackson-databind</artifactId>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>jakarta.annotation</groupId>
+      <artifactId>jakarta.annotation-api</artifactId>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>ch.qos.logback</groupId>
+      <artifactId>logback-classic</artifactId>
+      <scope>runtime</scope>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>io.micronaut.build</groupId>
+        <artifactId>micronaut-maven-plugin</artifactId>
+        <version>${micronaut-maven-plugin.version}</version>
+        <configuration>
+          <nativeImageBuildArgs>-Ob</nativeImageBuildArgs>
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <!-- Uncomment to enable incremental compilation -->
+          <!-- <useIncrementalCompilation>false</useIncrementalCompilation> -->
+          <annotationProcessorPaths combine.children="append">
+            <path>
+              <groupId>io.micronaut</groupId>
+              <artifactId>micronaut-graal</artifactId>
+              <version>${micronaut.version}</version>
+            </path>
+          </annotationProcessorPaths>
+          <compilerArgs>
+            <arg>-Amicronaut.processing.group=io.micronaut.build.examples</arg>
+            <arg>-Amicronaut.processing.module=${project.artifactId}</arg>
+          </compilerArgs>
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>com.google.cloud.tools</groupId>
+        <artifactId>jib-maven-plugin</artifactId>
+        <configuration>
+          <to>
+            <image>docker.io/alvarosanchez/demo:${project.version}</image>
+          </to>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+
+
+</project>

--- a/src/it/package-docker-native-different-parent-version/src/main/java/io/micronaut/build/examples/Application.java
+++ b/src/it/package-docker-native-different-parent-version/src/main/java/io/micronaut/build/examples/Application.java
@@ -1,0 +1,13 @@
+package io.micronaut.build.examples;
+
+import io.micronaut.runtime.Micronaut;
+import io.micronaut.build.examples.greeter.Greeter;
+
+public class Application {
+
+    public static void main(String[] args) {
+        Greeter greeter = new Greeter();
+        System.out.println(greeter.greet("world!"));
+        Micronaut.run(Application.class);
+    }
+}

--- a/src/it/package-docker-native-different-parent-version/src/main/java/io/micronaut/build/examples/greeter/Greeter.java
+++ b/src/it/package-docker-native-different-parent-version/src/main/java/io/micronaut/build/examples/greeter/Greeter.java
@@ -1,0 +1,8 @@
+package io.micronaut.build.examples.greeter;
+
+public class Greeter {
+
+    public String greet(String name) {
+        return "HELLO " + name;
+    }
+}

--- a/src/it/package-docker-native-different-parent-version/src/main/resources/application.yml
+++ b/src/it/package-docker-native-different-parent-version/src/main/resources/application.yml
@@ -1,0 +1,3 @@
+micronaut:
+  application:
+    name: package-docker-native

--- a/src/it/package-docker-native-different-parent-version/src/main/resources/application.yml
+++ b/src/it/package-docker-native-different-parent-version/src/main/resources/application.yml
@@ -1,3 +1,3 @@
 micronaut:
   application:
-    name: package-docker-native
+    name: package-docker-native-different-parent-version

--- a/src/it/package-docker-native-different-parent-version/src/main/resources/logback.xml
+++ b/src/it/package-docker-native-different-parent-version/src/main/resources/logback.xml
@@ -1,0 +1,15 @@
+<configuration>
+
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <withJansi>false</withJansi>
+        <!-- encoders are assigned the type
+             ch.qos.logback.classic.encoder.PatternLayoutEncoder by default -->
+        <encoder>
+            <pattern>%cyan(%d{HH:mm:ss.SSS}) %gray([%thread]) %highlight(%-5level) %magenta(%logger{36}) - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+</configuration>

--- a/src/it/package-docker-native-different-parent-version/src/test/java/io/micronaut/build/examples/ApplicationTest.java
+++ b/src/it/package-docker-native-different-parent-version/src/test/java/io/micronaut/build/examples/ApplicationTest.java
@@ -1,0 +1,21 @@
+package io.micronaut.build.examples;
+
+import io.micronaut.runtime.EmbeddedApplication;
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import javax.inject.Inject;
+
+@MicronautTest
+public class ApplicationTest {
+
+    @Inject
+    EmbeddedApplication application;
+
+    @Test
+    void testItWorks() {
+        Assertions.assertTrue(application.isRunning());
+    }
+
+}

--- a/src/it/package-docker-native-different-parent-version/verify.groovy
+++ b/src/it/package-docker-native-different-parent-version/verify.groovy
@@ -1,0 +1,4 @@
+File log = new File(basedir, 'build.log')
+assert log.exists()
+assert log.text.contains("BUILD FAILURE")
+assert log.text.contains("The io.micronaut:micronaut-parent version (3.4.0) differs from the micronaut.version property")

--- a/src/main/java/io/micronaut/build/DockerNativeMojo.java
+++ b/src/main/java/io/micronaut/build/DockerNativeMojo.java
@@ -51,6 +51,8 @@ public class DockerNativeMojo extends AbstractDockerMojo {
 
     public static final String DOCKER_NATIVE_PACKAGING = "docker-native";
     public static final String GRAALVM_ARGS = "GRAALVM_ARGS";
+    public static final String MICRONAUT_PARENT = "io.micronaut:micronaut-parent";
+    public static final String MICRONAUT_VERSION = "micronaut.version";
 
     @SuppressWarnings("CdiInjectionPointsInspection")
     @Inject
@@ -98,14 +100,29 @@ public class DockerNativeMojo extends AbstractDockerMojo {
     }
 
     private void checkGraalVm() throws MojoExecutionException {
-        String micronautVersion = mavenProject.getProperties().getProperty("micronaut.version");
-        if (!mavenProject.getInjectedProfileIds().get("io.micronaut:micronaut-parent:" + micronautVersion).contains("graalvm")) {
-            String javaVendor = System.getProperty("java.vendor", "");
-            if (javaVendor.toLowerCase().contains("graalvm")) {
-                throw new MojoExecutionException("The [graalvm] profile was not activated automatically because the native-image component is not installed (or not found in your path). Either activate the profile manually (-Pgraalvm) or install the native-image component (gu install native-image), and try again");
+        String micronautVersion = mavenProject.getProperties().getProperty(MICRONAUT_VERSION);
+        if (mavenProject.hasParent()) {
+            String ga = mavenProject.getParent().getGroupId() + ":" + mavenProject.getParent().getArtifactId();
+            if (MICRONAUT_PARENT.equals(ga)) {
+                String micronautParentVersion = mavenProject.getParent().getVersion();
+                if (micronautVersion.equals(micronautParentVersion)) {
+                    if (!mavenProject.getInjectedProfileIds().get("io.micronaut:micronaut-parent:" + micronautParentVersion).contains("graalvm")) {
+                        String javaVendor = System.getProperty("java.vendor", "");
+                        if (javaVendor.toLowerCase().contains("graalvm")) {
+                            throw new MojoExecutionException("The [graalvm] profile was not activated automatically because the native-image component is not installed (or not found in your path). Either activate the profile manually (-Pgraalvm) or install the native-image component (gu install native-image), and try again");
+                        } else {
+                            throw new MojoExecutionException("The [graalvm] profile was not activated automatically because you are not using a GraalVM JDK. Activate the profile manually (-Pgraalvm) and try again");
+                        }
+                    }
+                } else {
+                    String message = String.format("The %s version (%s) differs from the %s property (%s). Please, make sure both refer to the same version", MICRONAUT_PARENT, micronautParentVersion, MICRONAUT_VERSION, micronautVersion);
+                    throw new MojoExecutionException(message);
+                }
             } else {
-                throw new MojoExecutionException("The [graalvm] profile was not activated automatically because you are not using a GraalVM JDK. Activate the profile manually (-Pgraalvm) and try again");
+                getLog().warn("The parent POM of this project is not set to " + MICRONAUT_PARENT);
             }
+        } else {
+            getLog().warn("This project has no parent POM defined. To avoid build problems, please set the parent to " + MICRONAUT_PARENT);
         }
     }
 

